src/partition.jl

```diff
@@ -0,0 +1,150 @@
+mutable struct Partition
+    num_districts::Int
+    num_cut_edges::Int
+    total_population::Int
+
+    district_populations::Array{Int, 1} # of length(num_districts)
+    cut_edges::Array{Int, 1}            # of length(num_edges)
+    populations::Array{Int, 1}          # of length(num_nodes)
+    assignments::Array{Int, 1}          # of length(num_nodes)
+    district_adj::SparseMatrixCSC{Int, Int}
+    district_nodes::Array{Set{Int}, 1}
+    attributes::Array{Dict{String, Any}}
+end
+
+function Partition(graph::Graph,
+                   raw_graph::Dict{String, Any},
+                   pop_col::AbstractString,
+                   assignment_col::AbstractString,
+                   nodes_str::AbstractString = "nodes")::Partition
+    """ Builds the Partition Struct, which stores all the information of our
+        districting plan. It does not have any knowledge of the underlying graph
+        structure (that is available in the Graph object, which has reciprocally
+        has no idea of the districting plans.)
+
+        Arguments:
+            graph:     Graph object that has the underlying network structure of the plan.
+            raw_graph: Dict that was obtained by JSON.parsefile(filepath). Stores
+                       the attributes metadata (among others).
+            pop_col:   the key denoting the population attribute at the node level
+            assignment_col: the key denoting the district assignment at the node level
+            nodes_str: the key that has denotes the nodes
+    """
+    # get populations, assignments and num_districts
+    populations, assignments = get_populations_and_assignments(raw_graph, pop_col, assignment_col,
+                                                               graph.num_nodes, nodes_str)
+    num_districts = length(Set(assignments))
+
+    # get cut_edges, district_adjacencies
+    district_adj, cut_edges = get_district_adj_and_cut_edges(graph, assignments, num_districts)
+    num_cut_edges = sum(cut_edges)
+
+    # get district populations
+    district_populations = get_district_populations(assignments, populations, graph.num_nodes, num_districts)
+
+    # get total_population
+    total_population = sum(populations)
+
+    # get district_nodes
+    district_nodes = get_district_nodes(assignments, graph.num_nodes, num_districts)
+
+    # get attributes
+    attributes = get_attributes(raw_graph, graph.num_nodes, nodes_str)
+
+    return Partition(num_districts, num_cut_edges, total_population,
+                     district_populations, cut_edges, populations,
+                     assignments, district_adj, district_nodes, attributes)
+end
+
+################################################################################
```

Comment:
Just a personal preference, but I've always disliked separators like this—they make the code uglier without providing much semantic value. (I know these decorations are favored by some of the CS faculty here, though 😉)

Reply:
can you tell I went to Tufts CS?

Reply:
ill chuck it out